zstd/seqdec_amd64.s

```diff
@@ -0,0 +1,604 @@
+	//     llState = llTable[(llState.newState()+lowBits)&maxTableMask]
+	//     lowBits = uint16(bits >> (ofState.nbBits() & 31))
+	//     lowBits &= bitMask[mlState.nbBits()&15]
+	//     mlState = mlTable[(mlState.newState()+lowBits)&maxTableMask]
+	//     lowBits = uint16(bits) & bitMask[ofState.nbBits()&15]
+	//     ofState = ofTable[(ofState.newState()+lowBits)&maxTableMask]
+	// }
+	CMPQ decodeAsmContext_iteration(DI), $0
+	MOVQ ofState, R14                       // copy ofState, its current value is needed below
+	JZ   skip_update
+
+	MOVQ ctx+16(FP), R15
+
+	// lowBits := getBits(llState.nbBits())
+	MOVBQZX llState, AX // AX = nbBits() -- note: nbBits is the lowest byte of state
+
+	// BX = lowBits
+
+#ifdef GOAMD64_v3
+	LEAQ  (br_bits_read)(AX*1), CX
+	MOVQ  br_value, BX
+	MOVQ  CX, br_bits_read
+	ROLQ  CL, BX
+	BZHIQ AX, BX, BX
+
+#else
+	MOVQ    br_bits_read, CX // BX = ((br_value << br_bits_read) >> (64 - AX))
+	ADDQ    AX, br_bits_read
+	MOVQ    br_value, BX
+	SHLQ    CL, BX
+	MOVQ    AX, CX
+	XORQ    AX, AX
+	NEGQ    CX
+	SHRQ    CL, BX
+	TESTQ   CX, CX
+	CMOVQEQ AX, BX
+	MOVWQZX BX, BX
+
+#endif
+
+	MOVQ    llState, DX // DX = newState()
+	SHRQ    $16, DX
+	MOVWQZX DX, DX
+
+	ADDQ BX, DX            // DX += lowBits()
+	ANDQ $maxTableMask, DX // DX &= maxTableMask
+
+	// llState = llTable[...]
+	MOVQ decodeAsmContext_llTable(R15), AX
+	MOVQ (AX)(DX*8), llState
+
+	// lowBits := getBits(mlState.nbBits())
+	MOVBQZX mlState, AX // AX = nbBits() -- note: nbBits is the lowest byte of state
+
+	// BX = lowBits
+
+#ifdef GOAMD64_v3
+	LEAQ  (br_bits_read)(AX*1), CX
+	MOVQ  br_value, BX
+	MOVQ  CX, br_bits_read
+	ROLQ  CL, BX
+	BZHIQ AX, BX, BX
+
+#else
+	MOVQ    br_bits_read, CX // BX = ((br_value << br_bits_read) >> (64 - AX))
+	ADDQ    AX, br_bits_read
+	MOVQ    br_value, BX
+	SHLQ    CL, BX
+	MOVQ    AX, CX
+	XORQ    AX, AX
+	NEGQ    CX
+	SHRQ    CL, BX
+	TESTQ   CX, CX
+	CMOVQEQ AX, BX
+	MOVWQZX BX, BX
+
+#endif
+
+	MOVQ    mlState, DX // DX = newState()
+	SHRQ    $16, DX
+	MOVWQZX DX, DX
+
+	ADDQ BX, DX            // DX += lowBits()
+	ANDQ $maxTableMask, DX // DX &= maxTableMask
+
+	// mlState = llTable[...]
+	MOVQ decodeAsmContext_mlTable(R15), AX
+	MOVQ (AX)(DX*8), mlState
+
+	// lowBits := getBits(mlState.nbBits())
+	MOVBQZX ofState, AX // AX = nbBits() -- note: nbBits is the lowest byte of state
+
+	// BX = lowBits
+
+#ifdef GOAMD64_v3
+	LEAQ  (br_bits_read)(AX*1), CX
+	MOVQ  br_value, BX
+	MOVQ  CX, br_bits_read
+	ROLQ  CL, BX
+	BZHIQ AX, BX, BX
+
+#else
+	MOVQ    br_bits_read, CX // BX = ((br_value << br_bits_read) >> (64 - AX))
+	ADDQ    AX, br_bits_read
+	MOVQ    br_value, BX
+	SHLQ    CL, BX
+	MOVQ    AX, CX
+	XORQ    AX, AX
+	NEGQ    CX
+	SHRQ    CL, BX
+	TESTQ   CX, CX
+	CMOVQEQ AX, BX
+	MOVWQZX BX, BX
+
+#endif
+
+	MOVQ    ofState, DX // DX = newState()
+	SHRQ    $16, DX
+	MOVWQZX DX, DX
+
+	ADDQ BX, DX            // DX += lowBits()
+	ANDQ $maxTableMask, DX // DX &= maxTableMask
```

Comment:
I think this is just a bounds check elimination. It should be safe to remove.

Reply:
You are right. Tests pass after removal. Thank you.